x/ibc/core/02-client/abci.go

```diff
@@ -4,11 +4,32 @@ import (
 	sdk "github.com/cosmos/cosmos-sdk/types"
 	"github.com/cosmos/cosmos-sdk/x/ibc/core/02-client/keeper"
 	"github.com/cosmos/cosmos-sdk/x/ibc/core/exported"
+	ibctmtypes "github.com/cosmos/cosmos-sdk/x/ibc/light-clients/07-tendermint/types"
 )
 
 // BeginBlocker updates an existing localhost client with the latest block height.
 func BeginBlocker(ctx sdk.Context, k keeper.Keeper) {
-	_, found := k.GetClientState(ctx, exported.Localhost)
+	plan, found := k.GetUpgradePlan(ctx)
+	if found {
+		// Once we are at the last block this chain will commit, set the upgraded consensus state
+		// so that IBC clients can use the last NextValidatorsHash as a trusted kernel for verifying
+		// headers on the next version of the chain.
+		// Set the time to the last block time of the current chain.
+		// In order for a client to upgrade successfully, the first block of the new chain must be committed
+		// within the trusting period of the last block time on this chain.
+		_, exists := k.GetUpgradedClient(ctx, plan.Height)
```

Comment:
note this logic changed slightly. The previous logic used plan.IsIBCPlan() while this queries for the upgraded client state at `plan.Height`

Reply:
Hmm, what if there were two upgrades at the same height? I guess writing the consensus state twice is OK

Reply:
Two plans at the same height is strictly disallowed. The x/upgrade logic will always erase a previous plan if a new one is added. It is a contractual assumption that only at most one plan exists at a time. 

This does bring to mind that the following sequence is possible:
- create IBC upgrade at height x
- create new upgrade at height x + 1
- create new IBC upgrade at height x + 2

In this situation, the code will try to clear the IBC states at x + 1 since that is the last plan height, but the states exist at height x. I will add another key in the IBC store to store the latest IBC plan height and use this to clear the IBC states

Reply:
I updated this check to ensure that both a plan in x/upgrade is set and that it is an IBC upgrade